Multithreaded test execution/test groups

jstokes opened the pull request:
## Description

Provides a new option `:multithread` to run tests in parallel to address #22. By default, tests are run serially as they are today.

To ensure that tests that cannot be run concurrently don't overlap, this also allows tests to be grouped for execution using a `::test/group` metadata keyword. Tests that are in the same test group are executed in serial, with all test groups executing concurrently.

When executing tests in parallel all test output is buffered and then printed at the end of test execution to avoid interleaved messaging.

greglook commented on README.md:
```diff
@@ -9,7 +9,7 @@ a suite of tests against your systems gives you the confidence to _greenlight_
 
 - Steps should be _composable_ to drive down repetition in similar tests.
-- TODO: Tests should be _distributable_ and support parallelization out-of-the-box.
+- Tests should be _distributable_ and support parallelization out-of-the-box.
```
Hmm - parallelization is accurate, but technically we still haven't made tests distributable. Might be worth removing that word and adding a separate issue for distribution across multiple agents.

greglook commented on README.md:
```diff
@@ -328,6 +328,38 @@ to match test name.
 
+### Parallel Test Execution
+
+Tests can be executed in parallel by providing a `--multithread` option
```
Naming suggestion: maybe call this `--parallel` instead of `--multithread`? It's a bit easier to remember and more closely captures the intent instead of being tied to the implementation.

greglook commented on README.md:
```diff
@@ -328,6 +328,38 @@ to match test name.
+with a number of threads. Tests can be further grouped with `::test/group`
+metadata to indicate that tests within the same group should run serially.
+
+If a `::test/group` is not provided, a test is placed in its own group.
```
I really like this approach to the feature - it's a clever way of giving the user control without forcing serial tests to run after everything else.

Is there a way to select specific test groups to run? That could be a pretty nice feature for development as well as controlling some scenarios where you legitimately might want to pull out a test group for separate execution.

jstokes replied:
I extended the `runner/find-tests` function to optionally take a map of test properties that can be matched again - namespace, description or this test group.

greglook commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +81,60 @@
+         original-out# *out*
+         original-err# *err*]
+     (binding [*out* out#, *err* out#]
+       (let [result# ~@body]
```
This could expand into an invalid let binding since `body` is a varargs sequence. You need:
```suggestion
       (let [result# (do ~@body)]
```

greglook commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +81,60 @@
+       (let [result# ~@body]
+         (binding [*out* original-out#
+                   *err* original-err#]
+           (~printer (str out#))
```
Macro best practice: let `printer# ~printer` earlier in the macro, so that if someone passes in a form that _constructs_ a printer function, you aren't evaluating it inside these bindings.

jstokes replied:
I removed this delayed output printer for a piped reader/writer that allows streaming one test's output at a time, instead of waiting for test completion to print anything.

aengelberg commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +108,54 @@
+      (->> tests
+           (group-by #(or (::test/group %) (gensym)))
+           (map val)
+           (map (partial run-group exec-pool system))
```
I don't think that tests in the same linearized group also need to get piped to stdout adjacent to each other. I think it would make the most sense if tests get piped to stdout in the same order they were declared in `runner/main` (regardless of what group they're in). This would require some more async-fu to implement, but would gives the test author some extra control, if for example they know that some of their tests take especially long and don't want to block on those first.

Depending on how much you want to optimize this, you could also keep a queue of tests to pipe to stdout next, and bump things to the top of the queue if they fail in the background, which would surface errors sooner.

jstokes commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +108,54 @@
+  {:post [(instance? Reader (first %))
+          (instance? Future (second %))]}
+  (let [reader (PipedReader.)
+        writer (PipedWriter. reader)
```
Yeah, you're right. It looks like the default buffer size here is 1024 bytes, which does look like it ends up blocking test execution when attempting to print past that buffer length.

aria42 replied:
FWIW, you can choose a buffer size in the ctor of the class to be arbitrarily large.

jstokes replied:
Yeah, in at least the last iteration of using PipedReader/Writer I had this set to 1mb.

jstokes commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +108,54 @@
+    (loop []
+      (when-let [line (.readLine br)]
+        (println line)
+        (recur)))))
```
I'll pull this in if I can't get char-by-char working.

jstokes commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +108,54 @@
+                        (try
+                          (mapv (partial test/run-test! system) tests)
+                          (finally
+                            (.close writer)))))))]
```
I tried to make this `bound-fn` instead but was running into some issues, will revisit - may have not had the right arity.

jstokes commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +108,54 @@
+           (doall)
+           (mapcat identity))
+      (finally
+        (.shutdownNow exec-pool)))))
```
I don't think we're exposed to that since `step/advance!` is wrapped in a try/catch for catching test functions that happen to throw an Exception, to record an `:error` state.

aengelberg commented on src/greenlight/runner.clj:
```diff
@@ -112,27 +112,22 @@
-                            (.close writer)))))))]
+  (let [writer (PipedWriter.)
+        reader (PipedReader. writer (int (Math/pow 2 20)))
+        group-f (bound-fn
```
Does this need to be `bound-fn` as opposed to `fn`? I don't think you're looking up any vars that were bound outside of scope.

jstokes replied:
`test/*report*` is thread bound, yeah.

aengelberg replied:
Ah, carry on

jstokes commented:
Ok, removed the `PipedReader` and `PipedWriter` approach in favor of the simpler printing test output at the end of test execution. I added a bit of messaging around starting a test and also around the level of parallelization and number of test groups.

greglook commented on src/greenlight/runner.clj:
```diff
@@ -38,22 +44,41 @@
+    (filter #(re-matches matcher (name (:name (meta %)))))
+
+    true
+    (map (fn [v] (v)))
```
This part seems a little non-obvious to me - is it calling the test step constructor so that you can match against the produced attributes in the next test?

jstokes replied:
We were previously finding test vars and only filtering on metadata of those vars, then calling the test constructor after the filter. 

https://github.com/amperity/greenlight/blob/38898c149b3c723475d20a506546fd4984682a5f/src/greenlight/runner.clj#L53-L56

I wanted to extend that ability to also select on test properties (specifically the `::test/group`), but that requires calling the test constructor and then doing some additional filtering. 

I guess worst case we end up calling the test constructor (which could have side effects?) on tests that will not be selected.

greglook commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +102,68 @@
+        run-group (fn run-group
+                    [tests]
+                    (reify Callable
+                      (call [_]
```
I suspect you might just need the type-hint instead of this whole reification. The `submit` function takes both `Runbable` and `Callable<T>` in the single-argument signatures. Clojure functions implement both, for the zero-arity and one-arity signatures of `invoke`.

greglook commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +102,68 @@
+               [[_ group-tests]]
+               (.submit exec-pool ^Callable (run-group group-tests))))
+           (doall)
+           (map #(.get ^Future %))
```
Java futures support `IDeref`:
```suggestion
           (map deref)
```

greglook commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +102,68 @@
+           (doall)
+           (map #(.get ^Future %))
+           (doall)
+           (mapcat identity))
```
Could replace these two lines with:
```suggestion
           (into [] cat))
```

greglook commented on src/greenlight/runner.clj:
```diff
@@ -90,10 +177,12 @@
            (println "Running" (count tests) "tests...")
-           (let [results (mapv (partial test/run-test! system) tests)]
-             ; TODO: check result spec?
+           (let [results (if-let [n-threads (:parallel options)]
```
An idea: you could set a `:default 1` in the options definition, then test that it's greater than one before invoking the parallel path. You should definitely validate somewhere that if the option is set, it's a positive integer.

aengelberg commented on src/greenlight/runner.clj:
```diff
@@ -77,6 +102,68 @@
+         (binding [*out* original-out#
+                   *err* original-err#]
+           (printer# (str out#))
+           result#)))))
```
So stderr is effectively getting directed to stdout? That seems potentially expectation-breaking, but on the other hand, keeping two StringWriters and printing them both out at the end would result in reordered log lines, which could be even worse.

jstokes replied:
Yeah, I thought leaving the ordering the same was the lesser of two evils here.

greglook replied:
There is a much more complicated thing we could do here to preserve 'stderr-ness', but I think this is fine for now.

greglook commented on src/greenlight/runner.clj:
```diff
@@ -84,6 +110,64 @@
+     (binding [*out* out#, *err* out#]
+       (let [result# (do ~@body)]
+         (binding [*out* original-out#
+                   *err* original-err#]
```
A bit simpler way to do this:
```clojure
(let [out# (java.io.StringWriter.)
      printer# ~printer
      result# (binding [*out* out#
                        *err* out#]
                ~@body)]
  (printer# (str out#))
  result#)
```
```
